### ISSUE TITLE
Couldn't Kokkos::parallel_XXX return early when provided empty iteration spaces?

### ISSUE THREAD

**Issue by dutkalex** [kokkos/kokkos#6741]
Hi!
Using Kokkos on a daily basis over the past 3 years (btw thank you guys for the amazing job, I love Kokkos it really is a game changer IMHO), I have come to use all over my codebase the following pattern:
```cpp
template< size_t i >
void some_function( Kokkos::View< double*[i] > some_field ){
    if ( some_field.size() == 0 )
        return;

    Kokkos::parallel_for(
        "Parallel processing of the some_field data iff some_field is not empty",
        some_field.extent(0),
        KOKKOS_LAMBDA( size_t idx ){ /* do something smart */ }
    );
}
```
I started doing this because I needed to invoke ``some_funtion`` repeatedly with ``i = 0``, ``i = 1``, ..., ``i = N`` to cover for all the cases, while my users provide actual values only for a handful of values of ``i`` and all most of the kernels end up being conceptual no-ops in practice, and I noticed that adding this small early exit guard resulted in massive performance improvements due to lots of kernels not being launched at all. 
I am therefore wondering while does Kokkos::parrallel_for & co do not check for this. It seems to me like something that should be easy to check using the execution policy and would not induce any measurable overhead. Is it an oversight or is there a reason why such a thing would not be desirable?

**Comment by dutkalex** [comment]
I believe there was a significant performance overhead even with "simple" ``parallel_for`` loops, which constituted about 95% of the codebase at the time. Is it something that was implemented recently or has it always been like that from the beginning of Kokkos? I investigated this question during the spring/summer of 2022 and I was using Kokkos 3.something at the time...

**Comment by masterleinad** [comment]
We haven't audited all backends for the shortcut. Looking at `Cuda` quickly, it seems we are missing the shortcut only for `RangePolicy` `parallel_for`. It might be worth it to go through all combinations.

**Comment by dutkalex** [comment]
From an exterior perspective (i.e. feel free to disregard my comment since I don't know what I'm talking about 😉), it seems rather strange that this check is implemented Npatterns x Nbackends times. Can't it be implemented at the top level, before diving into the backend-related implementation details? I initially looked at the ``core/Kokkos_Parallel.hpp`` file, and since I did not find it there I assumed it was not implemented, hence my opening this issue...

**Comment by masterleinad** [comment]
Yes, it should be possible to do that for `parallel_for` at a higher level st least.
